homeassistant/components/zha/config_flow.py

```diff
@@ -53,48 +165,276 @@ async def async_step_user(self, user_input=None):
+                async with self._connect_zigpy_app() as app:
+                    await app.form_network()
+            elif strategy == FORMATION_REUSE_SETTINGS:
+                pass
+            elif strategy == FORMATION_RESTORE_MANUAL_BACKUP:
+                return await self.async_step_upload_manual_backup()
+            elif strategy == FORMATION_RESTORE_AUTOMATIC_BACKUP:
+                return await self.async_step_restore_automatic_backup()
+
+            return await self._async_create_radio_entity()
+
+        async with self._connect_zigpy_app() as app:
+            # Check if the stick has any settings and load them
+            try:
+                await app.load_network_info()
+            except NetworkNotFormed:
+                pass
+            else:
+                # Load the current info while we're connected, to save time
+                self._backups = app.backups.backups.copy()
+                self._current_settings = zigpy.backups.NetworkBackup(
+                    network_info=app.state.network_info,
+                    node_info=app.state.node_info,
+                )
+
+        strategies = []
+
+        # Check if we have any automatic backups *and* if the backups differ from
+        # the current radio settings (since restoring would be redundant)
+        if self._backups and any(
+            not backup.is_compatible_with(self._current_settings)
+            for backup in self._backups
+        ):
+            strategies.append(FORMATION_RESTORE_AUTOMATIC_BACKUP)
+
+        if self._current_settings is not None:
+            strategies.append(FORMATION_REUSE_SETTINGS)
+
+        strategies.append(FORMATION_RESTORE_MANUAL_BACKUP)
+        strategies.append(FORMATION_FORM_NEW_NETWORK)
+
+        # Pick a default. Forming a new network will always be an option but we would
+        # prefer to suggest something non-destructive if it's available.
+        suggested_strategy = [
+            strategy
+            for strategy in strategies
+            if strategy
+            in (
+                FORMATION_RESTORE_AUTOMATIC_BACKUP,
+                FORMATION_REUSE_SETTINGS,
+                FORMATION_FORM_NEW_NETWORK,
+            )
+        ][0]
+
+        schema = {
+            vol.Required(FORMATION_STRATEGY, default=suggested_strategy): vol.In(
+                strategies
+            ),
+        }
 
-        schema = {vol.Required(CONF_RADIO_TYPE): vol.In(sorted(RadioType.list()))}
         return self.async_show_form(
-            step_id="pick_radio",
+            step_id="choose_formation_strategy",
             data_schema=vol.Schema(schema),
         )
 
+    async def async_step_upload_manual_backup(self, user_input=None):
+        """Upload and restore a coordinator backup JSON file."""
+        errors = {}
+
+        if user_input is not None:
+            uploaded_file_id = user_input[UPLOADED_BACKUP_FILE]
+
+            with process_uploaded_file(self.hass, uploaded_file_id) as file_path:
```

Comment:
This context manager does I/O when exiting. We should schedule the whole call on the executor.

I'd extract the backup handling to a separate function that we can schedule on the executor from this step. Raise exceptions from that function to represent errors.